### DISCUSSION TITLE
how to add TraceId to response header

### DISCUSSION THREAD

**Discussion by Classical1956** [apache/skywalking#7644] [Q&A] [answered]
List of plugins used: 
![image](https://user-images.githubusercontent.com/13096704/131962157-c87337e5-8d70-4803-8ec0-0c958996c29f.png)

The project needs to return traceId in the response header, for which I used `OncePerRequestFilter` to achieve.
However, after starting the project according to the plugin-list above, only `TraceId=N/A` can be obtained.
After reviewing the documentation and skywalking project, add `tomcat-plugin` to the plugin-list. Although the traceId can be obtained, it causes new problems, like this issue [HTTP endpoints' module sometimes show as 'SpringMVC' but sometimes show as 'Tomcat Http'](https://github.com/apache/skywalking/issues/6905)

What do I need to do to get the traceId?

**Comment by wu-sheng** [comment, marked as answer]
We gave youmanual API toreadtrace id. Then, youcould do the rest by yourself.

**Reply by Classical1956** [reply]
is this `TraceContext.traceId()` ?

**Reply by wu-sheng** [reply]
Yes. Manipulating response header is not an option in SkyWalking core, so, you need to do that.
